crates/holochain/src/conductor/conductor.rs

```diff
@@ -3298,6 +3308,102 @@ mod countersigning_impls {
+            &self,
+            cell_id: &CellId,
+        ) -> ConductorResult<Option<CountersigningSessionState>> {
+            let space = self.get_or_create_space(cell_id.dna_hash())?;
+            let countersigning_workspaces = space.countersigning_workspaces.lock();
+            match countersigning_workspaces.get(cell_id) {
+                None => Ok(None),
+                Some(workspace) => Ok(workspace.get_countersigning_session_state()),
+            }
+        }
+
+        pub(crate) async fn abandon_countersigning_session(
+            &self,
+            cell_id: &CellId,
+        ) -> ConductorResult<()> {
+            let space = self.get_or_create_space(cell_id.dna_hash())?;
+            let (countersigning_workspace, preflight_request) = self
+                .get_preflight_request_from_unknown_session(&space, cell_id)
+                .await?;
+            force_abandon_session(space.clone(), cell_id.agent_pubkey(), &preflight_request)
+                .await?;
+            // Remove countersigning session from the workspace.
+            if countersigning_workspace
+                .remove_countersigning_session()
+                .is_none()
+            {
+                // The session exists in the space as previously checked, so this case must never happen.
+                tracing::warn!(
+                    ?cell_id,
+                    "Could not remove countersigning session from workspace after abandoning it."
+                );
+            }
+            Ok(())
+        }
+
+        pub(crate) async fn publish_countersigning_session(
+            &self,
+            cell_id: &CellId,
+        ) -> ConductorResult<()> {
+            let space = self.get_or_create_space(cell_id.dna_hash())?;
+            let (countersigning_workspace, preflight_request) = self
+                .get_preflight_request_from_unknown_session(&space, cell_id)
+                .await?;
+            let cell = self.cell_by_id(cell_id).await?;
+            let network = Arc::new(cell.holochain_p2p_dna().clone());
+            force_publish_countersigning_session(
+                space,
+                network,
+                self.keystore().clone(),
+                cell.triggers().integrate_dht_ops.clone(),
+                cell.triggers().publish_dht_ops.clone(),
+                cell_id.clone(),
+                preflight_request,
+            )
+            .await?;
+            if countersigning_workspace
+                .remove_countersigning_session()
+                .is_none()
+            {
+                // The session exists in the space as previously checked, so this case must never happen.
+                tracing::warn!(
+                    ?cell_id,
+                    "Could not remove countersigning session from workspace after publishing it."
+                );
+            }
+            Ok(())
+        }
+
+        async fn get_preflight_request_from_unknown_session(
```

Comment:
Also returns the workspace. Should it either not do that, or return the preflight request as the first item in the tuple, since that's what the function suggests is the important output